<!doctype html>
<html lang="en" prefix="og: http://ogp.me/ns#">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">

    <title>Manejo y transformación de errores en NodeJS</title>
    <meta name="description" content="Presentación de Manejo y Transformación de errores en NodeJS">

    <!-- Google Fonts -->
    <link href="https://fonts.googleapis.com/css?family=Roboto:100,100i,300,300i,400,400i,700,700i%7CMaitree:200,300,400,600,700&amp;subset=latin-ext" rel="stylesheet">

    <!-- CSS Base -->
    <link rel="stylesheet" type='text/css' media='all' href="static/css/webslides.css">

    <!-- Optional - CSS SVG Icons (Font Awesome) -->
    <link rel="stylesheet" type="text/css" media="all" href="static/css/svg-icons.css">
    <link rel="stylesheet" href="static/css/styles/default.css">
    <script src="static/js/highlight.pack.js"></script>
    <script>hljs.initHighlightingOnLoad();</script>

    <!-- Android -->
    <meta name="mobile-web-app-capable" content="yes">
    <meta name="theme-color" content="#333333">
  </head>
  <body>
    <header role="banner">
      <nav role="navigation">
        <ul>
          <li class="github">
            <a rel="external" href="https://nardhar.github.io/presentation-error-handling-nodejs/" title="Github">
              <svg class="fa-github">
                <use xlink:href="#fa-github"></use>
              </svg>
              <em>Custom Error Handling in NodeJS</em>
            </a>
          </li>
          <!--  <li class="dribbble"><a rel="external" href="http://dribbble.com/webslides" title="Dribbble"><svg class="fa-dribbble"><use xlink:href="#fa-dribbble"></use></svg> <em>webslides</em></a></li> -->
        </ul>
      </nav>
    </header>

    <main role="main">
      <article id="webslides">

        <section>
          <div class="wrap aligncenter">
            <h1>
              <strong>Manejando y Transformando errores con NodeJS</strong>
            </h1>
            <br>
            <br>
            <br>
            <h2>Felix Carreño</h2>
          </div>
        </section>

        <!-- Introduccion -->
        <section>
          <div class="wrap">
            <div class="grid vertical-align">
              <div class="column">
                <h3><strong>Erase una vez...</strong></h3>
              </div>
              <div class="column">
                <img src="static/images/cat-typing.gif" />
              </div>
            </div>
          </div>
        </section>

        <section>
          <div class="wrap">
            <h3><strong>Pero...</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <img src="static/images/javascript-stacktrace-1.png" />
              </div>
              <div class="column">
                <img src="static/images/cat-cry.png" />
              </div>
            </div>
          </div>
        </section>

        <section>
          <div class="wrap">
            <h3><strong>También...</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <img src="static/images/hold-a-teacup.png" />
              </div>
              <div class="column">
                <img src="static/images/user-bad-tea.png" />
              </div>
            </div>
          </div>
        </section>

        <section>
          <div class="wrap">
            <h3><strong>Y en otros casos...</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <img src="static/images/system-architecture.png" />
              </div>
              <div class="column">
                <img src="static/images/chrome.jpg" />
              </div>
            </div>
          </div>
        </section>

        <!-- 1.2 Manejo de errores (try/catch) -->
        <section>
          <div class="wrap">
            <h3><strong>Usando try/catch</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">
try {
  console.log(item.length);
} catch(e) {
  console.log("ERROR: " + e.message);
}
// imprime: ERROR: item is not defined
                  </code>
                </pre>
              </div>
              <div class="column">
                <pre>
                  <code class="javascript">function f(item) {
  try {
    var len = item.length;
    try {
      return item.split(' ');
    } catch(e) {
      console.log("ERROR: item no es un array");
    }
  } catch(e) {
    console.log("ERROR: " + e.message);
  }
}
f(2);
// ERROR: item no es un array</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <!-- 2.1 Funciones como Ciudadanos de primera categoria -->
        <section>
          <div class="wrap">
            <h3><strong>Funciones como <em>"First class citizens"</em> 1/3</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">function esNumeroPar(num) {
  return num % 2 === 0;
}

var listado = [1,4,8,51,90,101];
console.log(listado.filter(esNumeroPar));
// filter es una función de orden superior (high order function)

// imprime: [4, 8, 90]</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <section>
          <div class="wrap">
            <h3><strong>Funciones como <em>"First class citizens"</em> 2/3</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">function esNumeroPar(num) {
  return num % 2 === 0;
}
function esMultiplo3(num) {
  return num % 3 === 0;
}
function esMultiplo4(num) {
  return num % 4 === 0;
}
...
</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <section>
          <div class="wrap">
            <h3><strong>Funciones como <em>"First class citizens"</em> 3/3</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">function esMultiploDe(divisor) {
  return function multiplo(num) {
    return num % divisor === 0;
  };
}

var listado = [1,4,8,51,90,101];
console.log(listado.filter(esMultiploDe(2))); // [4, 8, 90]
console.log(listado.filter(esMultiploDe(3))); // [51, 90]
console.log(listado.filter(esMultiploDe(4))); // [4, 8]</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <!-- 1.3 Extendiendo la clase Error (ES6) -->
        <section>
          <div class="wrap">
            <h3><strong>La clase Error de NodeJS</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                Todos los errores en NodeJS extienden de la clase <strong>Error</strong>
                <p>
                  <ul>
                    <li>- TypeError</li>
                    <li>- ReferenceError</li>
                    <li>- SyntaxError</li>
                    <li>- ...</li>
                  </ul>
                </p>
              </div>
              <div class="column">
                <pre>
                  <code class="javascript">class ApiError extends Error {
  constructor(message, status) {
    super(message);
    this.status = status;
  }
}
class ValidationError extends ApiError {
  constructor(message, fieldErrors) {
    super(message, 412);
    this.fieldErrors = fieldErrors;
  }
}</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <!-- 1.4 Ejemplo con throw -->
        <section>
          <div class="wrap">
            <h3><strong>Ventaja de Extender clases</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">function validarUsuario(datosUsuario) {
  // ...
  throw new ValidationError("Error al validar el usuario", []);
}
try {
  validarUsuario({ nombre: 'Felix' });
} catch(e) {
  console.log(e instanceof ValidationError); // true
  console.log(e instanceof ApiError); // true
}</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <!-- 3.1 Propagando errores (async/await y promesas) -->
        <section class="slide-top">
          <div class="wrap">
            <h3><strong>Propagando errores</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">function validarUsuario(datosUsuario) {
  return metodoAsincrono(datosUsuario)
  .then((usuario) => {
    throw new ValidationError('Rol', {
      nombre: datosUsuario.rol,
    });
  });
}
function registrarUsuario(datosUsuario) {
  return validarUsuario(datosUsuario)
  .catch((e) => {
    if (e instanceof ValidationError) {
      console.log(
        e.constructor.name,
        e.fieldErrors
      );
    } else {
      console.log(e);
    }
  });
}
registrarUsuario({});
// ValidationError { nombre: undefined }</code>
                </pre>
              </div>
              <div class="column">
                <pre>
                  <code class="javascript">async function validarUsuario(datosUsuario) {
  try {
    const usuario = await metodoAsincrono(datosUsuario);
    throw new ValidationError('Rol', {
      nombre: datosUsuario.rol,
    });
  } catch(error) { throw error; }
}
async function registrarUsuario(datosUsuario) {
  try {
    validarUsuario(datosUsuario);
  } catch(e) {
    if (e instanceof ValidationError)
      console.log(
        e.constructor.name,
        e.fieldErrors
      );
    else console.log(e);
  });
}
registrarUsuario({});
// ValidationError { nombre: undefined }</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <!-- 3.2 Transformando errores (async/await y promesas) -->
        <section class="slide-top">
          <div class="wrap">
            <h3><strong>Transformando errores 1/3</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">function validarUsuario(datosUsuario) {
  return buscarRol(datosUsuario.rol)
  .catch((e) => { // si falla la ejecución
    if (e instanceof DatabaseError) {
      throw new NotFoundError('Rol', {
        nombre: datosUsuario.rol,
      });
    }
    throw e;
  });
}
function registrarUsuario(datosUsuario) {
  return validarUsuario(datosUsuario)
  .catch((e) => {
    if (e instanceof NotFoundError) {
      return null;
    }
    throw e;
  });
}
registrarUsuario({});
// null</code>
                </pre>
              </div>
              <div class="column">
                <pre>
                  <code class="javascript">async function validarUsuario(datosUsuario) {
  try {
    await buscarRol(datosUsuario.rol);
  } catch(error) {
    if (e instanceof DatabaseError) {
      throw new NotFoundError('Rol', {
        nombre: datosUsuario.rol,
      });
    }
    throw e;
  }
}
async function registrarUsuario(datosUsuario) {
  try {
    validarUsuario(datosUsuario);
  } catch(e) {
    if (e instanceof NotFoundError) {
      return null;
    }
    throw e;
  });
}
registrarUsuario({});
// null</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <section class="slide-top">
          <div class="wrap">
            <h3><strong>Transformando errores 2/3 (agregando manejadores)</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">function handleDatabaseError(callback) {
  return function handler(err) {
    if (err instanceof DatabaseError) {
      if (typeof callback === 'function') {
        return callback(err);
      }
      return callback;
    }
  };
}
function handleNotFoundError(callback) {
  return function handler(err) {
    if (err instanceof NotFoundError) {
      if (typeof callback === 'function') {
        return callback(err);
      }
      return callback;
    }
  };
}</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <section class="slide-top">
          <div class="wrap">
            <h3><strong>Transformando errores 3/3</strong></h3>
            <div class="grid vertical-align">
              <div class="column">
                <pre>
                  <code class="javascript">function validarUsuario(datosUsuario) {
  return buscarRol(datosUsuario.rol)
  .catch(
    handleDatabaseError((e) => {
      throw new NotFoundError('Rol', {
        nombre: datosUsuario.rol,
      })
    })
  );
}
function registrarUsuario(datosUsuario) {
  return validarUsuario(datosUsuario)
  .catch(handleNotFoundError(null));
}
registrarUsuario({});
// null</code>
                </pre>
              </div>
              <div class="column">
                <pre>
                  <code class="javascript">function validarUsuario(datosUsuario) {
  try {
    return buscarRol(datosUsuario.rol);
  } catch(e) {
    return handleDatabaseError((e2) => {
      throw new NotFoundError('Rol', {
        nombre: datosUsuario.rol,
      })
    });
  }
}
function registrarUsuario(datosUsuario) {
  try {
    return validarUsuario(datosUsuario);
  } catch(e) {
    return handleNotFoundError(null)(e);
  }
}
registrarUsuario({});
// null</code>
                </pre>
              </div>
            </div>
          </div>
        </section>

        <section class="slide-top">
          <div class="wrap">
            <h3><strong>Ejemplo 1:</strong></h3>
            <div class="">
              <pre>
                <code class="javascript">return usuarioService.verificarContrasena({
  usuario: params.usuario,
  contrasena: encrypt(params.contrasena),
}) // si es exitoso retorna el usuario
.catch(NotFoundError.handle(null))
.then((usuario) => {
  // guarda el log de registro
  return registroService.guardarIngreso(params.usuario)
  .catch(ValidationError.handle(() => {
    throw new NotFoundError('Usuario', [{ field: 'usuario', code: 'invalid' }]);
  }))
  .then(() => {
    return usuario;
  });
});</code>
              </pre>
            </div>
          </div>
        </section>

        <section class="slide-top">
          <div class="wrap">
            <h3><strong>Ejemplo 2:</strong></h3>
            <div class="content-center">
              <p>
                <ul>
                  <li>
                    <a href="https://github.com/nardhar/node-backend-skel/">https://github.com/nardhar/node-backend-skel/</a>
                  </li>
                  <li>
                    <a href="https://github.com/nardhar/mini-mvcs/">https://github.com/nardhar/mini-mvcs/</a>
                  </li>
                </ul>
              </p>
            </div>
          </div>
        </section>

        <!-- 5 Gracias -->
        <section class="aligncenter">
          <h3><strong>¡Gracias!</strong></h3>
          <hr>
          <p>
            <svg class="fa-link">
              <use xlink:href="#fa-link"></use>
            </svg>
            <a href="https://github.com/nardhar/presentation-error-handling-nodejs" title="Github link">https://github.com/nardhar/presentation-error-handling-nodejs</a>
          </p>
          <p>
            <svg class="fa-github">
              <use xlink:href="#fa-github"></use>
            </svg>
            <strong>Github</strong> <a href="https://github.com/nardhar" title="Github account">@nardhar</a>
          </p>
          <p>
            <svg class="fa-twitter">
              <use xlink:href="#fa-twitter"></use>
            </svg>
            <strong>Twitter</strong> <a href="https://twitter.com/nardharmure" title="Twitter account">@nardharmure</a>
          </p>
        </section>
      </article>
    </main>
    <!--main-->

    <!-- Required -->
    <script src="static/js/webslides.js"></script>

    <script>
      window.ws = new WebSlides();
    </script>

    <!-- OPTIONAL - svg-icons.js (fontastic.me - Font Awesome as svg icons) -->
    <script defer src="static/js/svg-icons.js"></script>

  </body>
</html>
